### PR TITLE
Tests on Makefile

### PR DESCRIPTION
**Pull request by ayushkovskiy**
Currently, lacks only test on `make train` (see https://github.com/neuromation/cookiecutter-neuro-project/issues/39) and local make-operations (`setup-local`,`lint`,`install`,`ps`).

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ jobs:
       # Download and cache dependencies
       - restore_cache:
           keys:
-            - v1-dependencies-{{ checksum "requirements-dev.txt" }}
+            - &pip_packages_cache_key v1-dependencies-3.6.1-{{ checksum "requirements-dev.txt" }}
```
Comment (ayushkovskiy) [inline .circleci/config.yml]: better syntax

File: {{cookiecutter.project_slug}}/Makefile [hunk 2]
```diff
@@ -48,12 +50,14 @@ setup: ### Setup remote environment
 		--volume $(PROJECT_PATH_STORAGE):$(PROJECT_PATH_ENV):ro \
 		$(BASE_ENV_NAME) \
 		'sleep 1h'
+	@echo
```
Comment (ayushkovskiy) [inline {{cookiecutter.project_slug}}/Makefile]: needed for testing: somehow no space was printed

File: {{cookiecutter.project_slug}}/Makefile [hunk 3]
```diff
@@ -48,12 +50,14 @@ setup: ### Setup remote environment
...
 	neuro cp $(REQUIREMENTS_APT) $(REQUIREMENTS_APT_STORAGE)
 	neuro cp $(REQUIREMENTS_PIP) $(REQUIREMENTS_PIP_STORAGE)
-	neuro exec $(SETUP_NAME) "bash -c 'apt-get update && cat $(REQUIREMENTS_APT_ENV) | xargs -I % apt-get install -y --no-install-recommends % && apt-get clean && apt-get autoremove && rm -rf /var/lib/apt/lists/*'"
-	neuro exec $(SETUP_NAME) 'pip install -r $(REQUIREMENTS_PIP_ENV)'
+	neuro exec --no-key-check $(SETUP_NAME) "bash -c 'export DEBIAN_FRONTEND=noninteractive && $(APT_COMMAND) update && cat $(REQUIREMENTS_APT_ENV) | xargs -I % $(APT_COMMAND) install --no-install-recommends % && $(APT_COMMAND) clean && $(APT_COMMAND) autoremove && rm -rf /var/lib/apt/lists/* && echo APT requirements installation completed'"
+	neuro exec --no-key-check $(SETUP_NAME) "bash -c '$(PIP_INSTALL_COMMAND) -r $(REQUIREMENTS_PIP_ENV) && echo PIP requirements installation completed'"
 	neuro --network-timeout 300 job save $(SETUP_NAME) $(CUSTOM_ENV_NAME)
 	neuro kill $(SETUP_NAME)
+	@echo Project setup completed
```
Comment (ayushkovskiy) [inline {{cookiecutter.project_slug}}/Makefile]: also optional, may be deleted. But we check it in tests to make them more robust.

File: {{cookiecutter.project_slug}}/Makefile [hunk 4]
```diff
@@ -79,7 +83,7 @@ upload-notebooks:  ### Upload notebooks directory to Storage
 
 .PHONY: download-notebooks
 download-notebooks:  ### Download notebooks directory from Storage
-	neuro cp -r $(NOTEBOOKS_PATH_STORAGE) $(NOTEBOOKS_PATH)
+	neuro cp -r -T $(NOTEBOOKS_PATH_STORAGE) $(NOTEBOOKS_PATH)
```
Comment (ayushkovskiy) [inline {{cookiecutter.project_slug}}/Makefile]: it's a bug

File: {{cookiecutter.project_slug}}/Makefile [hunk 5]
```diff
@@ -48,12 +50,14 @@ setup: ### Setup remote environment
 		--volume $(PROJECT_PATH_STORAGE):$(PROJECT_PATH_ENV):ro \
 		$(BASE_ENV_NAME) \
 		'sleep 1h'
+	@echo
 	neuro cp $(REQUIREMENTS_APT) $(REQUIREMENTS_APT_STORAGE)
 	neuro cp $(REQUIREMENTS_PIP) $(REQUIREMENTS_PIP_STORAGE)
-	neuro exec $(SETUP_NAME) "bash -c 'apt-get update && cat $(REQUIREMENTS_APT_ENV) | xargs -I % apt-get install -y --no-install-recommends % && apt-get clean && apt-get autoremove && rm -rf /var/lib/apt/lists/*'"
-	neuro exec $(SETUP_NAME) 'pip install -r $(REQUIREMENTS_PIP_ENV)'
+	neuro exec --no-key-check $(SETUP_NAME) "bash -c 'export DEBIAN_FRONTEND=noninteractive && $(APT_COMMAND) update && cat $(REQUIREMENTS_APT_ENV) | xargs -I % $(APT_COMMAND) install --no-install-recommends % && $(APT_COMMAND) clean && $(APT_COMMAND) autoremove && rm -rf /var/lib/apt/lists/* && echo APT requirements installation completed'"
```
Comment (ayushkovskiy) [inline {{cookiecutter.project_slug}}/Makefile]: I know, it's, better do changes in Makefile in a separate PR

File: cleanup_jobs.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            cmd = f"neuro status {job}"
+            out = run_command(cmd, detect_new_jobs=False)
+            print(f"`{cmd}` => {repr(out)}")
+
+            cmd = f"neuro kill {job}"
+            out = run_command(cmd, detect_new_jobs=False)
+            print(f"`{cmd}` => {repr(out)}")
+            killed.append(job)
+
+        except Exception as e:
```
Comment (ayushkovskiy) [inline cleanup_jobs.py]: I know, bad design, but this is a helper script. To be cleaned up later.

File: tests/e2e/test_e2e_template.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+            debug=True,
+            timeout=TIMEOUT_MAKE_SETUP,
+            expect_patterns=expected_patterns,
+            # TODO: add specific error patterns
+            stop_patterns=DEFAULT_ERROR_PATTERNS,
+        )
+
+
+@pytest.mark.run(order=2)
+def test_make_upload_clean_code() -> None:
```
Comment (ayushkovskiy) [inline tests/e2e/test_e2e_template.py]: this test flakes sometimes: https://circleci.com/gh/neuromation/cookiecutter-neuro-project/106

Comment (ayushkovskiy) [reply tests/e2e/test_e2e_template.py]: Let's put `@flake(retries=3)` on it?

Comment (ayushkovskiy) [reply tests/e2e/test_e2e_template.py]: added global option `--reruns=2`: we all know how e2e tests flake

Comment (ayushkovskiy) [reply tests/e2e/test_e2e_template.py]: this means each failed test will run once again. Usually it helps

File: {{cookiecutter.project_slug}}/Makefile [hunk 8]
```diff
@@ -107,16 +122,17 @@ training:  ### Run training job
...
 connect-training:  ### Execute shell to the training job
-	neuro exec $(TRAINING_NAME) bash
+	neuro exec --no-key-check $(TRAINING_NAME) bash
 
 .PHONY: jupyter
 jupyter:  ### Run jupyter job
 	neuro run \
 		--name $(JUPYTER_NAME) \
-		--preset gpu-small \
+		--preset cpu-small \
```
Comment (ayushkovskiy) [inline {{cookiecutter.project_slug}}/Makefile]: Temporary change: currently, a `gpu-small` job cannot start within 10 minutes on `staging`. 
cc @mariyadavydova

Comment (mariyadavydova) [reply {{cookiecutter.project_slug}}/Makefile]: We can't merge this branch into master then.
I would better add `jupyter-cpu` (see #37) and test that command.

File: {{cookiecutter.project_slug}}/Makefile (hunk 4 above)
Comment (mariyadavydova) [reply {{cookiecutter.project_slug}}/Makefile]: True. I've created a related issue (#45).

File: {{cookiecutter.project_slug}}/Makefile (hunk 2 above)
Comment (mariyadavydova) [reply {{cookiecutter.project_slug}}/Makefile]: I don't like having these echos in our production Makefile. Do we really need them?

Comment (ayushkovskiy) [reply {{cookiecutter.project_slug}}/Makefile]: no, i implemented run helpers in a way that we really don't need these echoes. Maria please delete them (also from the rest code)

Comment (ayushkovskiy) [reply {{cookiecutter.project_slug}}/Makefile]: wait: this is another kind of echo: it's needed to add a new line. try to delete or and room tests, of they fail, please fix the test patterns.

Comment (ayushkovskiy) [reply {{cookiecutter.project_slug}}/Makefile]: so other echoes like "echo Success" can be deleted

File: {{cookiecutter.project_slug}}/Makefile (hunk 8 above)
Comment (ayushkovskiy) [reply {{cookiecutter.project_slug}}/Makefile]: please discuss with the team how we can solve the problem that gpu jobs are not always available. If it's unsolvable, @anaiden knows the hack how to allow circle ci to run jobs that don't print anything for more that 10 min

File: {{cookiecutter.project_slug}}/Makefile (hunk 4 above)
Comment (ayushkovskiy) [reply {{cookiecutter.project_slug}}/Makefile]: thanks. let's then fix makefile first, and then rebase this branch

File: {{cookiecutter.project_slug}}/Makefile [hunk 9]
```diff
@@ -133,10 +133,9 @@ connect-training:  ### Execute shell to the training job
 jupyter: upload-code upload-notebooks ### Run jupyter job
 	neuro run \
 		--name $(JUPYTER_NAME) \
-
 		--preset $(TRAINING_MACHINE_TYPE) \
 		--http 8888 --detach \
-		$(HTTP_AUTH) \
+    $(HTTP_AUTH) \
```
Comment (mariyadavydova) [inline {{cookiecutter.project_slug}}/Makefile]: There is something wrong with tabs here, it won't work.
